### ISSUE TITLE
Baron adds newline at end of file

### ISSUE THREAD

**Issue by DRMacIver** [Psycojoker/baron#70]
This is an incredibly pedantic bug that you probably don't care about, but it appears to be the case that baron fails to satisfy the invariant dumps(parse(s)) == s when s does not have a newline at the end of the file because baron adds one to the result of dumps even if it was not present in the source.

**Comment by Psycojoker** [comment]
Yes, it's a very specific situation when the endl token is included in indented code, you can find the details here: https://github.com/Psycojoker/baron/blob/master/baron/baron.py#L38-43

It's fixable but it's very complex for a very small thing (you have to do a recursive traversal of the tree to get to the last ever displayed token which is the endl one, then remove it, hoping not to break things).

**Comment by adamchainz** [comment]
Technically a text file isn't a text file under posix unless it finishes with `\n` ( https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline )

**Comment by Psycojoker** [comment]
While this isn't my priority right now, I think this situation is valid when you try to work on snippets of code outside of a file (but I have yet to find someone that has published code doing this for now).

**Comment by ibizaman** [comment]
This could be solved by handling partial trees. Although that's far in the future.
